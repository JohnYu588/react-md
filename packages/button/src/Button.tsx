import {
  RippleContainer,
  useElementInteraction,
  useHigherContrastChildren,
} from "@react-md/core";
import type { ButtonHTMLAttributes } from "react";
import { forwardRef } from "react";
import type { ButtonClassNameThemeOptions } from "./styles";
import { getButtonClassName } from "./styles";

export interface ButtonProps
  extends ButtonHTMLAttributes<HTMLButtonElement>,
    ButtonClassNameThemeOptions {
  /** @defaultValue `"button"` */
  type?: "button" | "reset" | "submit";
}

/**
 * @example
 * Simple Example
 * ```tsx
 * import { Button } from "@react-md/button";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       Content
 *     </Button>
 *   );
 * }
 * ```
 *
 * @example
 * Theme Example
 * ```tsx
 * import { Button } from "@react-md/button";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       theme="primary"
 *       themeType="contained"
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       Content
 *     </Button>
 *   );
 * }
 * ```
 *
 * @example
 * Icon Button Example
 * ```tsx
 * import { Button } from "@react-md/button";
 * import { FavoriteSVGIcon } from "@react-md/material-icons";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       theme="secondary"
 *       themeType="outline"
 *       buttonType="icon"
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       <FavoriteSVGIcon />
 *     </Button>
 *   );
 * }
 * ```
 *
 * @example
 * Text Button with icons
 * ```tsx
 * import { Button } from "@react-md/button";
 * import { FavoriteSVGIcon } from "@react-md/material-icons";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       <FavoriteSVGIcon />
 *       Content
 *     </Button>
 *   );
 * }
 * ```
 */
export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  function Button(props, ref) {
    const {
      type = "button",
      disabled = false,
      theme = "clear",
      themeType = "flat",
      buttonType = "text",
      className,
      children: propChildren,
      onClick,
      onKeyDown,
      onKeyUp,
      onMouseDown,
      onMouseUp,
      onMouseLeave,
      onTouchStart,
      onTouchEnd,
      onTouchMove,
      ...remaining
    } = props;
    const isThemeDisabled = theme === "disabled";
    const { pressed, pressedClassName, rippleContainerProps, handlers } =
      useElementInteraction({
        onClick,
        onKeyDown,
        onKeyUp,
        onMouseDown,
        onMouseUp,
        onMouseLeave,
        onTouchStart,
        onTouchEnd,
        onTouchMove,
        disabled: disabled || isThemeDisabled,
      });

    const children = useHigherContrastChildren(propChildren);

    return (
      <button
        {...remaining}
        // when the theme is set to `"disabled"`, the event handlers should be
        // removed so that it behaves like a disabled button. you do not want to
        // actually set the `disabled` attribute since it will lose keyboard
        // focus. this is mostly for supporting circular progress bars within
        // buttons
        {...(isThemeDisabled ? {} : handlers)}
        aria-disabled={isThemeDisabled || undefined}
        ref={ref}
        type={type}
        className={getButtonClassName({
          theme,
          themeType,
          buttonType,
          disabled,
          pressed,
          pressedClassName,
          className,
        })}
      >
        {children}
        {rippleContainerProps && <RippleContainer {...rippleContainerProps} />}
      </button>
    );
  }
);
