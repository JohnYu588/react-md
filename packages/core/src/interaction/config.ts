import { type ElementInteractionMode } from "./types.js";

/**
 * @remarks \@since 6.0.0
 */
export interface ElementInteractionConfiguration {
  mode: ElementInteractionMode;
  higherContrast: boolean;
}

// NOTE: Uses get/set for test mocking

let mode: ElementInteractionMode = "ripple";
let higherContrast = true;

/**
 * @remarks \@since 6.0.0
 */
export const INTERACTION_CONFIG: ElementInteractionConfiguration = {
  get mode() {
    return mode;
  },
  set mode(nextMode: ElementInteractionMode) {
    mode = nextMode;
  },
  get higherContrast() {
    return higherContrast;
  },
  set higherContrast(nextHigherContrast) {
    higherContrast = nextHigherContrast;
  },
};
