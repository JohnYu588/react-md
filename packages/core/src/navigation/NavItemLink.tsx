"use client";
import {
  forwardRef,
  useEffect,
  type AnchorHTMLAttributes,
  type LiHTMLAttributes,
} from "react";
import { cssUtils } from "../cssUtils.js";
import { useElementInteraction } from "../interaction/useElementInteraction.js";
import { Tooltip } from "../tooltip/Tooltip.js";
import { useTooltip } from "../tooltip/useTooltip.js";
import { type PropsWithRef } from "../types.js";
import { useEnsuredRef } from "../useEnsuredRef.js";
import { NavItem } from "./NavItem.js";
import { navItemLink } from "./navItemStyles.js";
import {
  type NavItemContentProps,
  type NavigationLinkComponent,
} from "./types.js";

/**
 * @since 6.0.0
 */
export interface NavItemLinkProps
  extends Omit<AnchorHTMLAttributes<HTMLAnchorElement>, "children">,
    NavItemContentProps {
  /** @defaultValue `"a"` */
  as?: NavigationLinkComponent;
  href: string;

  /**
   * Any additional props to provide the wrapping `<li>` element such as
   * `style`, `className`, and `ref`.
   */
  liProps?: PropsWithRef<LiHTMLAttributes<HTMLLIElement>, HTMLLIElement>;

  /**
   * Set this to `true` if the link matches the current `pathname`. This will
   * enable `aria-current="true"` and apply active styling.
   */
  active: boolean;

  /** @defaultValue `cssUtils({ fontWeight: "bold" })` */
  activeClassName?: string;

  /** @defaultValue `!to && !href` */
  disabled?: boolean;

  /**
   * Set this to `true` to prevent this item from scrolling into view when it
   * becomes active. It is generally recommended to keep this `true` unless a
   * custom implementation has been added so that when the temporary navigation
   * will show the active route.
   *
   * @defaultValue `false`
   */
  disableScrollIntoView?: boolean;
}

/**
 * **Client Component**
 *
 * @since 6.0.0
 */
export const NavItemLink = forwardRef<HTMLAnchorElement, NavItemLinkProps>(
  function NavItemLink(props, propRef) {
    const {
      as: Component = "a",
      href,
      active,
      activeClassName,
      beforeAddon,
      afterAddon,
      className,
      children,
      liProps,
      spanProps,
      disabled = !href,
      onClick,
      onBlur,
      onDragStart,
      onKeyDown,
      onKeyUp,
      onMouseDown,
      onMouseLeave,
      onMouseUp,
      onTouchMove,
      onTouchEnd,
      onTouchStart,
      tooltipOptions,
      disableScrollIntoView,
      ...remaining
    } = props;

    const [linkRef, linkRefCallback] = useEnsuredRef(propRef);
    const { handlers, ripples } = useElementInteraction({
      disabled,
      onClick,
      onBlur,
      onDragStart,
      onKeyDown,
      onKeyUp,
      onMouseDown,
      onMouseLeave,
      onMouseUp,
      onTouchMove,
      onTouchEnd,
      onTouchStart,
    });

    const { elementProps, tooltipProps, overflowRef } = useTooltip({
      overflowOnly: true,
      defaultPosition: "right",
      ...tooltipOptions,
      ...handlers,
    });
    useEffect(() => {
      const link = linkRef.current;
      if (
        !active ||
        !link ||
        disableScrollIntoView ||
        document.activeElement === link
      ) {
        return;
      }

      link.scrollIntoView({ block: "center" });
    }, [active, disableScrollIntoView, linkRef]);

    return (
      <NavItem {...liProps}>
        <Component
          {...remaining}
          {...handlers}
          {...elementProps}
          ref={linkRefCallback}
          href={href}
          className={navItemLink({
            active,
            activeClassName,
            className,
          })}
        >
          {beforeAddon}
          <span
            {...spanProps}
            ref={overflowRef}
            className={cssUtils({
              className: spanProps?.className,
              textOverflow: "ellipsis",
            })}
          >
            {children}
          </span>
          {afterAddon}
          {ripples}
        </Component>
        <Tooltip {...tooltipProps}>{children}</Tooltip>
      </NavItem>
    );
  }
);
