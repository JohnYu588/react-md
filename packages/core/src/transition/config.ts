/**
 * @since 6.0.0
 */
export interface TransitionConfig {
  /**
   * Set this to `true` to disable all transitions from `react-md`.
   *
   * Note: It is recommended to set this to `true` in testing to keep things
   * simple and will automatically be set when using:
   *
   * ```ts
   * import "@react-md/core/test-utils/jest-globals/setup";
   *
   * // or
   * import "@react-md/core/test-utils/vitest/setup";
   * ```
   *
   * @defaultValue `false`
   */
  disabled: boolean;
}

// NOTE: Uses get/set for test mocking

let disabled = false;

/**
 * @since 6.0.0
 */
export const TRANSITION_CONFIG: TransitionConfig = {
  get disabled() {
    return disabled;
  },
  set disabled(nextDisabled: boolean) {
    disabled = nextDisabled;
  },
};
