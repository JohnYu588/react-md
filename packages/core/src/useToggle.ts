"use client";

import { useCallback, useState } from "react";

import { type UseStateInitializer, type UseStateSetter } from "./types.js";

/**
 * @since 6.0.0
 */
export interface ToggleImplementation {
  toggled: boolean;
  setToggled: UseStateSetter<boolean>;
  toggle: () => void;
  enable: () => void;
  disable: () => void;
}

/**
 *
 * @example Simple Example
 * ```tsx
 * import { Button, useToggle } from "@react-md/core";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   const { toggled, toggle } = useToggle();
 *
 *   return (
 *     <>
 *       <Button onClick={toggle}>Toggle</Button>
 *       {`Toggled: ${toggled}`}
 *     </>
 *   );
 * }
 * ```
 *
 * @param defaultValue - `false`
 */
export function useToggle(
  defaultValue: UseStateInitializer<boolean> = false
): ToggleImplementation {
  const [toggled, setToggled] = useState(defaultValue);

  return {
    toggled,
    setToggled,
    toggle: useCallback(() => {
      setToggled((prevToggled) => !prevToggled);
    }, []),
    enable: useCallback(() => {
      setToggled(true);
    }, []),
    disable: useCallback(() => {
      setToggled(false);
    }, []),
  };
}
