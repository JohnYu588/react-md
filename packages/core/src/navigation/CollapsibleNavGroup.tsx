"use client";
import { forwardRef, type LiHTMLAttributes, type ReactNode } from "react";
import { type ButtonProps } from "../button/Button.js";
import {
  type CollapseTransitionHookOptions,
  useCollapseTransition,
} from "../transition/useCollapseTransition.js";
import { type PropsWithRef } from "../types.js";
import { NavGroup, type NavGroupProps } from "./NavGroup.js";
import { NavItem } from "./NavItem.js";
import {
  NavItemButton,
  type NavItemButtonRotatorProps,
} from "./NavItemButton.js";

const noop = (): void => {
  // do nothing
};

/**
 * @since 6.0.0
 */
export interface CollapsibleNavGroupProps
  extends NavGroupProps,
    NavItemButtonRotatorProps {
  liProps?: PropsWithRef<LiHTMLAttributes<HTMLLIElement>, HTMLLIElement>;
  buttonProps?: PropsWithRef<ButtonProps, HTMLButtonElement>;
  collapseOptions?: Omit<
    CollapseTransitionHookOptions<HTMLUListElement>,
    "nodeRef" | "className" | "transitionIn"
  >;

  /**
   * The children to render in the {@link NavItemButton} that toggles the
   * collapsed state
   */
  buttonChildren: ReactNode;

  /**
   * The content that should be rendered in the non-collapsed {@link NavGroup}.
   */
  children: ReactNode;

  collapsed: boolean;
  toggleCollapsed: () => void;
}

/**
 * **Client Component**
 *
 * @since 6.0.0
 */
export const CollapsibleNavGroup = forwardRef<
  HTMLUListElement,
  CollapsibleNavGroupProps
>(function CollapsibleNavGroup(props, ref) {
  const {
    liProps,
    buttonProps,
    buttonChildren,
    icon,
    iconRotatorProps,
    disableIconRotator,
    className,
    children,
    collapsed,
    toggleCollapsed,
    collapseOptions,
    ...remaining
  } = props;

  const { rendered, elementProps } = useCollapseTransition({
    ...collapseOptions,
    nodeRef: ref,
    className,
    transitionIn: !collapsed,
  });
  const onButtonClick = buttonProps?.onClick ?? noop;

  return (
    <NavItem {...liProps}>
      <NavItemButton
        {...buttonProps}
        onClick={(event) => {
          onButtonClick(event);
          toggleCollapsed();
        }}
        collapsed={collapsed}
        icon={icon}
        iconRotatorProps={iconRotatorProps}
        disableIconRotator={disableIconRotator}
      >
        {buttonChildren}
      </NavItemButton>
      {rendered && (
        <NavGroup {...remaining} {...elementProps}>
          {children}
        </NavGroup>
      )}
    </NavItem>
  );
});
