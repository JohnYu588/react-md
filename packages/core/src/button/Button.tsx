"use client";
import { cnb } from "cnbuilder";
import type { ButtonHTMLAttributes } from "react";
import { forwardRef } from "react";
import { RippleContainer } from "../interaction/RippleContainer";
import { useElementInteraction } from "../interaction/useElementInteraction";
import { useHigherContrastChildren } from "../interaction/useHigherContrastChildren";
import type { PropsWithRef } from "../types";
import { bem } from "../utils";
import type {
  FloatingActionButtonPosition,
  FloatingActionButtonProps,
} from "./FloatingActionButton";
import { FloatingActionButton } from "./FloatingActionButton";

declare module "react" {
  interface CSSProperties {
    "--rmd-button-background-color"?: string;
    "--rmd-button-color"?: string;
    "--rmd-button-border-radius"?: string | number;
    "--rmd-button-contained-background-color"?: string;
    "--rmd-button-contained-color"?: string;
    "--rmd-button-text-horizontal-padding"?: string | number;
    "--rmd-button-text-vertical-padding"?: string | number;
    "--rmd-button-text-min-height"?: string | number;
    "--rmd-button-text-min-width"?: string | number;
    "--rmd-button-icon-size"?: string | number;
    "--rmd-button-icon-radius"?: string | number;
  }
}

const styles = bem("rmd-button");

/**
 * When this is set to `"text"`, the size of the button will be determined by
 * the content and will be more block-like. Icons can still be rendered in
 * `"text"` buttons and will have spacing automatically applied between other
 * content in the button.
 *
 * When this is set to `"icon"`, the button will be equal height/width and
 * circular.
 */
export type ButtonType = "text" | "icon" | "icon-square";

/**
 * One of the valid material design default button themes that can be used. This
 * will update the general look and feel by updating the colors within the
 * button while the `ButtonThemeType` will update the borders or box shadow.
 */
export type ButtonTheme =
  | "clear"
  | "primary"
  | "secondary"
  | "warning"
  | "error"
  | "disabled";

/**
 * When this is set to `"flat"`, the button will have no `background-color`,
 * `border`, and `box-shadow`. It will only set the `color` to the
 * {@link ButtonTheme}.
 *
 * When this is set to `"outline"`, the button will have no `background-color`,
 * but gain a `border` and `color` set to the {@link ButtonTheme}.
 *
 * When this is set to `"contained"`, the button will set the `background-color`
 * to the {@link ButtonTheme}, add some `box-shadow`, and set the `color` to
 * either `#000` or `#fff`. (The `color` defaults to whichever value has the
 * highest contrast ratio with the `background-color`)
 */
export type ButtonThemeType = "flat" | "outline" | "contained";

/** @remarks \@since 6.0.0 */
export interface ButtonClassNameThemeOptions {
  className?: string;

  /** @defaultValue `false` */
  disabled?: boolean;

  /**
   * @see {@link ButtonTheme} for information about the different types.
   * @defaultValue `"text"`
   */
  buttonType?: ButtonType;

  /** @defaultValue `"clear"` */
  theme?: ButtonTheme;

  /**
   * @see {@link ButtonThemeType} for information about the theming behavior.
   * @defaultValue `"flat"`
   */
  themeType?: ButtonThemeType;
}

/**
 * @remarks \@since 6.0.0
 * @internal
 */
export interface ButtonClassNameOptions extends ButtonClassNameThemeOptions {
  /** @defaultValue `false` */
  pressed?: boolean;
  pressedClassName?: string;
}

/**
 * Creates a button theme based on the button theming props. This is really just
 * used so that other elements like clickable `<div>`s or `<input type="file">`
 * can look like buttons.
 *
 * @param options - An object containing the themeable button props to generate a
 * button theme className.
 * @returns a string of class names to create an element with a button theme.
 * @remarks \@since 6.0.0 This used to be called `buttonThemeClassNames`.
 */
export function button(options: ButtonClassNameOptions = {}): string {
  const {
    theme: propTheme = "clear",
    themeType = "flat",
    buttonType = "text",
    disabled: propDisabled = false,
    pressed = false,
    pressedClassName,
    className,
  } = options;

  const theme = propTheme === "disabled" ? "clear" : propTheme;
  const disabled = propDisabled || propTheme === "disabled";
  const text = buttonType === "text";
  const icon = !text;
  const outline = themeType === "outline";
  const contained = themeType === "contained";
  const clear = theme === "clear";

  return cnb(
    styles({
      text,
      icon,
      "icon-square": buttonType === "icon-square",
      disabled,
      contained: !disabled && contained,
      outline,
      pressed: contained && pressed,
      [theme]: !disabled && !clear && contained,
      [`text-${theme}`]: !disabled && !clear && !contained,
      [`outline-${theme}`]: !disabled && !clear && outline,
    }),
    pressedClassName,
    className
  );
}

export interface ButtonProps
  extends ButtonHTMLAttributes<HTMLButtonElement>,
    ButtonClassNameThemeOptions {
  /** @defaultValue `"button"` */
  type?: "button" | "reset" | "submit";

  /**
   * The position within the viewport to display the button as a floating action
   * button.
   */
  floating?: FloatingActionButtonPosition;

  /**
   * Any additional props to provide the to `FAB` container element when the
   * `floating` prop is provided
   */
  floatingProps?: PropsWithRef<FloatingActionButtonProps, HTMLSpanElement>;
}

/**
 * **Client Component**
 *
 * @example
 * Simple Example
 * ```tsx
 * import { Button } from "@react-md/core";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       Content
 *     </Button>
 *   );
 * }
 * ```
 *
 * @example
 * Theme Example
 * ```tsx
 * import { Button } from "@react-md/core";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       theme="primary"
 *       themeType="contained"
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       Content
 *     </Button>
 *   );
 * }
 * ```
 *
 * @example
 * Icon Button Example
 * ```tsx
 * import { Button } from "@react-md/core";
 * import { FavoriteSVGIcon } from "@react-md/material-icons";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       theme="secondary"
 *       themeType="outline"
 *       buttonType="icon"
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       <FavoriteSVGIcon />
 *     </Button>
 *   );
 * }
 * ```
 *
 * @example
 * Text Button with icons
 * ```tsx
 * import { Button } from "@react-md/core";
 * import { FavoriteSVGIcon } from "@react-md/material-icons";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       <FavoriteSVGIcon />
 *       Content
 *     </Button>
 *   );
 * }
 * ```
 */
export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  function Button(props, ref) {
    const {
      type = "button",
      disabled = false,
      floating = null,
      floatingProps,
      theme = floating ? "secondary" : "clear",
      themeType = floating ? "contained" : "flat",
      buttonType = floating ? "icon" : "text",
      className,
      children: propChildren,
      onBlur,
      onClick,
      onKeyDown,
      onKeyUp,
      onMouseDown,
      onMouseUp,
      onMouseLeave,
      onDragStart,
      onTouchStart,
      onTouchEnd,
      onTouchMove,
      ...remaining
    } = props;
    const isThemeDisabled = theme === "disabled";
    const { pressed, pressedClassName, rippleContainerProps, handlers } =
      useElementInteraction({
        onBlur,
        onClick,
        onKeyDown,
        onKeyUp,
        onMouseDown,
        onMouseUp,
        onMouseLeave,
        onDragStart,
        onTouchStart,
        onTouchEnd,
        onTouchMove,
        disabled: disabled || isThemeDisabled,
      });

    const children = useHigherContrastChildren(propChildren);

    return (
      <FloatingActionButton position={floating} {...floatingProps}>
        <button
          {...remaining}
          // when the theme is set to `"disabled"`, the event handlers should be
          // removed so that it behaves like a disabled button. you do not want to
          // actually set the `disabled` attribute since it will lose keyboard
          // focus. this is mostly for supporting circular progress bars within
          // buttons
          {...(isThemeDisabled ? {} : handlers)}
          aria-disabled={isThemeDisabled || remaining["aria-disabled"]}
          disabled={disabled}
          ref={ref}
          type={type}
          className={button({
            theme,
            themeType,
            buttonType,
            disabled,
            pressed,
            pressedClassName,
            className,
          })}
        >
          {children}
          {rippleContainerProps && (
            <RippleContainer {...rippleContainerProps} />
          )}
        </button>
      </FloatingActionButton>
    );
  }
);
