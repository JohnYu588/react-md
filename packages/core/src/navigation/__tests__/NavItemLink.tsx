import { beforeEach, describe, expect, it, jest } from "@jest/globals";
import { type ReactElement, createRef } from "react";

import { Button } from "../../button/Button.js";
import {
  render,
  screen,
  userEvent,
  waitFor,
  waitForElementToBeRemoved,
} from "../../test-utils/index.js";
import { TRANSITION_CONFIG } from "../../transition/config.js";
import { useToggle } from "../../useToggle.js";
import { CollapsibleNavGroup } from "../CollapsibleNavGroup.js";
import { NavItemLink } from "../NavItemLink.js";

const scrollIntoView = jest.spyOn(
  HTMLAnchorElement.prototype,
  "scrollIntoView"
);

beforeEach(() => {
  jest.resetAllMocks();
});

describe("NavItemLink", () => {
  it("should apply the correct styling, HTML attributes, and allow a ref", () => {
    const ref = createRef<HTMLAnchorElement>();
    const props = {
      ref,
      href: "/",
      active: false,
      children: "Home",
    } as const;
    const { rerender } = render(<NavItemLink {...props} />);
    const navItem = screen.getByRole("listitem");
    const link = screen.getByRole("link", { name: "Home" });

    expect(ref.current).toBeInstanceOf(HTMLAnchorElement);
    expect(ref.current).toBe(link);
    expect(navItem).toMatchSnapshot();

    rerender(
      <NavItemLink
        {...props}
        style={{ color: "white" }}
        className="custom-class-name"
      />
    );
    expect(navItem).toMatchSnapshot();

    rerender(<NavItemLink {...props} active />);
    expect(navItem).toMatchSnapshot();

    rerender(
      <NavItemLink {...props} active activeClassName="custom-active-class" />
    );
    expect(navItem).toMatchSnapshot();

    rerender(<NavItemLink {...props} activeClassName="custom-active-class" />);
    expect(navItem).toMatchSnapshot();
  });

  it("should support an overflow only tooltip by default", async () => {
    const user = userEvent.setup();
    render(
      <NavItemLink href="/" active spanProps={{ "data-testid": "content" }}>
        Home
      </NavItemLink>
    );
    const link = screen.getByRole("link", { name: "Home" });

    const content = screen.getByTestId("content");
    jest.spyOn(content, "scrollWidth", "get").mockReturnValue(300);

    const rect = content.getBoundingClientRect();
    jest.spyOn(content, "offsetWidth", "get").mockReturnValue(150);
    jest
      .spyOn(content, "getBoundingClientRect")
      .mockReturnValue({ ...rect, width: 150 });

    await user.hover(link);
    const tooltip = await screen.findByRole("tooltip", {
      name: "Home",
    });

    await waitFor(() => {
      expect(tooltip).toHaveStyle("position: fixed; left: 24px; top: 16px;");
    });
    expect(tooltip).toMatchSnapshot();
  });

  it("should allow an addon to be rendered before or after the children so that the addons are not included in the overflowing tooltip", () => {
    render(
      <NavItemLink
        href="/"
        active
        beforeAddon={<span>Before</span>}
        afterAddon={<span>After</span>}
      >
        Home
      </NavItemLink>
    );

    const navItem = screen.getByRole("listitem");
    expect(navItem).toMatchSnapshot();
  });

  it("should automatically scroll the item into view when it gains the active state to better support temporary layouts", async () => {
    const user = userEvent.setup();
    function Test(): ReactElement {
      const { toggled, toggle } = useToggle();

      return (
        <>
          <Button onClick={toggle}>Toggle</Button>
          <NavItemLink href="/" active={toggled}>
            Home
          </NavItemLink>
        </>
      );
    }

    render(<Test />);
    const toggle = screen.getByRole("button", { name: "Toggle" });

    await user.click(toggle);
    expect(scrollIntoView).toHaveBeenCalledWith({ block: "center" });

    await user.click(toggle);
    expect(scrollIntoView).toHaveBeenCalledTimes(1);

    await user.click(toggle);
    expect(scrollIntoView).toHaveBeenCalledTimes(2);
  });

  it("should not scroll the item into view if it became active by the user clicking on the link", async () => {
    const user = userEvent.setup();
    function Test() {
      const { toggle, toggled } = useToggle();

      return (
        <NavItemLink href="/" active={toggled} onClick={toggle}>
          Home
        </NavItemLink>
      );
    }

    render(<Test />);
    const link = screen.getByRole("link", { name: "Home" });
    expect(link).not.toHaveAttribute("aria-current");

    await user.click(link);
    expect(link).toHaveAttribute("aria-current", "page");
    expect(scrollIntoView).not.toHaveBeenCalled();
  });

  it("should not scroll the item into view if the disableScrollIntoView prop is enabled", async () => {
    const user = userEvent.setup();
    function Test(): ReactElement {
      const { toggled, toggle } = useToggle();

      return (
        <>
          <Button onClick={toggle}>Toggle</Button>
          <NavItemLink href="/" active={toggled} disableScrollIntoView>
            Home
          </NavItemLink>
        </>
      );
    }

    render(<Test />);
    const toggle = screen.getByRole("button", { name: "Toggle" });

    await user.click(toggle);
    expect(scrollIntoView).not.toHaveBeenCalled();
  });

  it("should not scroll the item into view if it became active by mounting after a collapse transition since it makes it animate weirdly", async () => {
    TRANSITION_CONFIG.disabled = false;
    const user = userEvent.setup();
    function Test(): ReactElement {
      const { toggle, toggled } = useToggle(true);
      return (
        <>
          <CollapsibleNavGroup
            depth={0}
            collapsed={toggled}
            buttonChildren="Group"
            toggleCollapsed={toggle}
          >
            <NavItemLink href="/" active>
              Home
            </NavItemLink>
          </CollapsibleNavGroup>
        </>
      );
    }
    render(<Test />);

    const toggle = screen.getByRole("button", { name: "Group" });
    expect(() => screen.getByRole("link")).toThrow();

    await user.click(toggle);
    const link = await screen.findByRole("link", { name: "Home" });
    await waitFor(() => {
      expect(document.querySelector(".rmd-collapse--enter")).toBeNull();
    });
    expect(scrollIntoView).not.toHaveBeenCalled();

    await user.click(toggle);
    await waitForElementToBeRemoved(link);
    expect(scrollIntoView).not.toHaveBeenCalled();

    // same thing, shouldn't scroll into view if transitions are disabled
    TRANSITION_CONFIG.disabled = true;
    await user.click(toggle);
    await screen.findByRole("link", { name: "Home" });
    expect(document.querySelector(".rmd-collapse--enter")).toBeNull();
    expect(scrollIntoView).not.toHaveBeenCalled();
  });
});
