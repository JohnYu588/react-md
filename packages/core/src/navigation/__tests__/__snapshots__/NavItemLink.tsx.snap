// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavItemLink should allow an addon to be rendered before or after the children so that the addons are not included in the overflowing tooltip 1`] = `
<li
  class="rmd-nav-item"
>
  <a
    aria-current="page"
    class="rmd-nav-item__content rmd-nav-item__link rmd-nav-item__link--active rmd-button rmd-button--text rmd-interaction-surface rmd-none rmd-bold"
    href="/"
    id=":r4:"
  >
    <span>
      Before
    </span>
    <span
      class="rmd-nowrap rmd-ellipsis"
    >
      Home
    </span>
    <span>
      After
    </span>
  </a>
</li>
`;

exports[`NavItemLink should apply the correct styling, HTML attributes, and allow a ref 1`] = `
<li
  class="rmd-nav-item"
>
  <a
    class="rmd-nav-item__content rmd-nav-item__link rmd-button rmd-button--text rmd-interaction-surface rmd-none"
    href="/"
    id=":r0:"
  >
    <span
      class="rmd-nowrap rmd-ellipsis"
    >
      Home
    </span>
  </a>
</li>
`;

exports[`NavItemLink should apply the correct styling, HTML attributes, and allow a ref 2`] = `
<li
  class="rmd-nav-item"
>
  <a
    class="rmd-nav-item__content rmd-nav-item__link rmd-button rmd-button--text rmd-interaction-surface rmd-none custom-class-name"
    href="/"
    id=":r0:"
    style="color: white;"
  >
    <span
      class="rmd-nowrap rmd-ellipsis"
    >
      Home
    </span>
  </a>
</li>
`;

exports[`NavItemLink should apply the correct styling, HTML attributes, and allow a ref 3`] = `
<li
  class="rmd-nav-item"
>
  <a
    aria-current="page"
    class="rmd-nav-item__content rmd-nav-item__link rmd-nav-item__link--active rmd-button rmd-button--text rmd-interaction-surface rmd-none rmd-bold"
    href="/"
    id=":r0:"
    style=""
  >
    <span
      class="rmd-nowrap rmd-ellipsis"
    >
      Home
    </span>
  </a>
</li>
`;

exports[`NavItemLink should apply the correct styling, HTML attributes, and allow a ref 4`] = `
<li
  class="rmd-nav-item"
>
  <a
    aria-current="page"
    class="rmd-nav-item__content rmd-nav-item__link rmd-nav-item__link--active rmd-button rmd-button--text rmd-interaction-surface rmd-none custom-active-class"
    href="/"
    id=":r0:"
    style=""
  >
    <span
      class="rmd-nowrap rmd-ellipsis"
    >
      Home
    </span>
  </a>
</li>
`;

exports[`NavItemLink should apply the correct styling, HTML attributes, and allow a ref 5`] = `
<li
  class="rmd-nav-item"
>
  <a
    class="rmd-nav-item__content rmd-nav-item__link rmd-button rmd-button--text rmd-interaction-surface rmd-none"
    href="/"
    id=":r0:"
    style=""
  >
    <span
      class="rmd-nowrap rmd-ellipsis"
    >
      Home
    </span>
  </a>
</li>
`;

exports[`NavItemLink should support an overflow only tooltip by default 1`] = `
<span
  class="rmd-tooltip rmd-tooltip--right rmd-tooltip--visible"
  id=":r2:-tooltip"
  role="tooltip"
  style="position: fixed; left: 24px; top: 16px;"
>
  Home
</span>
`;
